devict/voteict
# Message tagging, and subscription management

There's a lot going on here, but from a high level..

- Subscribers will now be able to sign in to the system using just their phone number by receiving a verification code.
- Both Subscribers and ScheduledMessages have tags now (locations and topics).
  - Admin can manage these tags.
  - Tags can be set as "subscriber" or "message" defaults, meaning new messages/subscribers will have them by default.
- A subscriber can control what tags they are subscribed to from a simple web UI, and enable/disable subscription.
- When a message goes out, a subscriber must have one overlapping location tag and one overlapping topic tag to receive it.

Since promoting the service by phone number has never really been great, with this new web UI for sign up and subscription management, we can just promote the URL!

### How to test this out

Start the system locally, make sure twilio integration is working. Run a fresh migration and DB seeding.

```
$ docker-compose run --rm php php artisan migrate:fresh
$ make db-seed
```

**To test out subscriber management..**

- Start login by clicking "Get started now!" on the homepage.
  - Perform the login flow, entering the verification code.
- From the subscriber dashboard..
  - Attempt some tag management.
  - Attempt enabling/disabling subscription, verify in DB.

**To test the scheduled message tagging..**

- Be a subscriber from your phone number, have at least one tag you're subscribed to and one you aren't.
- Visit /login, use admin@example.com // Password!1
- Create a new scheduled message tagged with the one you're subscribed to, set to deliver in a minute or so.
- Repeat with a tag that you're not subscribed to.
- After the timestamps have passed, run `make docker-process-messages` to send off the messages.
  - If not running in docker, use `php artisan schedule:run`.

**Generally just poke around a lot and see if anything seems broken, or if anything could be a bit better.**
